Add support for references to CiliumCIDRGroup inside FromCIDRSet for ingress rules in CNPs

pippolo84 opened the pull request:
Add support for `cidrGroupRef` in CIDR ingress rules in Cilium Network Policies.

The new rule allows to reference a list of CiliumCIDRGroup objects that contain the CIDRs from which we want to allow traffic from.

An example of the newly introduced CiliumCIDRGroup CRD is the following:

```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumCIDRGroup
metadata:
  name: test-group-1
spec:
  externalCIDRs:
    - "172.19.0.1/32"
```

The feature is implemented as a wrapper to the existing FromCIDR Ingress rule. Each `cidrGroupRef` field is translated to the equivalent `fromCIDRSet` block with the CIDRs from the referenced CiliumCIDRGroup object.

To do that, we use a single handler for CiliumNetworkPolicy, CiliumClusterwideNetworkPolicy and CiliumCIDRGroup in the k8s watcher.
    
Every time the referenced CiliumCIDRGroups are updated, the CNP rules are recalculated, in order to append to the Ingress.FromCIDR field the updated lists of CIDRs.
    
In case the lists of CIDRs in the referenced CiliumCIDRGroups are all empty, the policy will have an Ingress rule with all fields equal to nil, except the `cidrGroupRef` one.  That would result in the same behavior as an empty FromCIDR rule, which means default deny.

This PR also adds some metrics to get info about the new Ingress rule impact and usage. Specifically:

- number of events related to CiliumCIDRGroup received and processed
- time taken to translate the `cidrGroupRef` field into CIDRs
- number of CNPs and CCNPs that are referencing at least a CiliumCIDRGroup object

Fixes #10349

```
Add new CRD CiliumCIDRGroup which allows users to define external CIDRs to be used within a CiliumNetworkPolicy, outside of the CiliumNetworkPolicy, via fromCIDRSet using cidrGroupRef
```

pippolo84 committed:
CRDs: Add CiliumCIDRGroup CRD and scaffolding

This generates the CiliumCIDRGroup type, to reference a group of CIDRs
as a single entity in CiliumNetworkPolicies.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
k8s: Add support for cidrGroupRef field in CNP/CCNP

Add support for cidrGroupRef field in Ingress rules for Cilium Network
Policies.  The new field allows to reference a list of CiliumCIDRGroup
objects that contain the CIDRs from which we wants to allow traffic
from.

The feature is implemented as a wrapper to the existing FromCIDRSet
Ingress rule. Each policy with a reference to a cidr group is translated
in an equivalent "FromCIDRSet policy" after looking up the CIDRs in the
referenced CiliumCIDRGroup objects.

To do that, we use a single handler for CiliumNetworkPolicy,
CiliumClusterwideNetworkPolicy and CiliumCIDRGroup in the k8s watcher,
in order to avoid further locking when accessing the CNP/CCNP cache or
the CIDRGroup one.

Every time the referenced CiliumCIDRGroups are updated, the CNP rules
are recalculated, in order to append to the Ingress.FromCIDRSet field
the updated lists of CIDRs.

In case the referenced CiliumCIDRGroups k8s objects are all empty or
deleted, the policy behaviour is default-deny.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

pippolo84 committed:
k8s: Unify CNP and CCNP handling in watchers

CNP and CCNP events handling is based on the same code, since each CNP
and CCNP is first translated to an intermediate types.SlimCNP
representation. To support the handling of CNP/CCNP updates, there are
two separate caches, one for CNP and one for CCNP.

This commit unify the events handling and the caches, relying on the two
event channels to avoid any additional locking mechanism.

This sets the stage for the fromCIDRGroupRef feature, where we need to
reference another k8s object (CiliumCIDRGroup) to get the CIDRs the
policies are going to allow when filtering ingress traffic.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -388,3 +566,130 @@ func reportCNPChangeMetrics(err error) {
+			}
+		}
+	}
+	for i := range translated.Specs {
```
Why aren't we checking for `translated.Specs` if it's nil like we do for `Spec`?

pippolo84 replied:
If `translated.Specs` is `nil`, we won't enter the for-range loop and no panic will happen.

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -160,22 +244,24 @@ func (k *K8sWatcher) onUpsertCNP(
 	// See https://github.com/cilium/cilium/blob/27fee207f5422c95479422162e9ea0d2f2b6c770/pkg/policy/api/ingress.go#L112-L134
 	cnpCpy := cnp.DeepCopy()
 
+	translatedCNP := resolveCIDRGroupRef(cnp, cidrGroupCache)
```
```suggestion
	translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
```

This function should use the copy and that way, it doesn't have to make a copy inside the function.

pippolo84 replied:
I made it this way because I need to store a copy of the cnp in the cache ahead of the translation. So, in case of a cidr group ref being present, I need to make another copy (the previous one will be stored in the cache and the translated one will be consumed/modified by the add/update cnp code)

pippolo84 replied:
But I agree that it should use `cnpCpy` anyway, that it's a copy-paste error.

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +277,98 @@ func (k *K8sWatcher) onDeleteCNP(
+		// See https://github.com/cilium/cilium/blob/27fee207f5422c95479422162e9ea0d2f2b6c770/pkg/policy/api/ingress.go#L112-L134
+		cnpCpy := cnp.DeepCopy()
+
+		translatedCNP := resolveCIDRGroupRef(cnp, cidrGroupCache)
```
Again, this should be using the `cnpCpy`

```suggestion
		translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
```

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +277,98 @@ func (k *K8sWatcher) onDeleteCNP(
+
+		translatedCNP := resolveCIDRGroupRef(cnp, cidrGroupCache)
+
+		err := k.updateCiliumNetworkPolicyV2(cs, cnp, translatedCNP, initialRecvTime)
```
```suggestion
		err := k.updateCiliumNetworkPolicyV2(cs, cnpCpy, translatedCNP, initialRecvTime)
```

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +277,98 @@ func (k *K8sWatcher) onDeleteCNP(
+		// See https://github.com/cilium/cilium/blob/27fee207f5422c95479422162e9ea0d2f2b6c770/pkg/policy/api/ingress.go#L112-L134
+		cnpCpy := cnp.DeepCopy()
+
+		translatedCNP := resolveCIDRGroupRef(cnp, cidrGroupCache)
```
Use the copy here as well

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +277,98 @@ func (k *K8sWatcher) onDeleteCNP(
+
+		translatedCNP := resolveCIDRGroupRef(cnp, cidrGroupCache)
+
+		err := k.updateCiliumNetworkPolicyV2(cs, cnp, translatedCNP, initialRecvTime)
```
Use the copy here as well

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -388,3 +566,130 @@ func reportCNPChangeMetrics(err error) {
+}
+
+func translateFromCIDRGroupRefs(cnp *types.SlimCNP, cidrsSets map[string][]api.CIDR) *types.SlimCNP {
+	cnpCpy := cnp.DeepCopy()
```
If we use the copies correctly, then I believe we won't need this

pippolo84 replied:
I think we do, to keep the "untranslated" cnp in the cache. See [this comment](https://github.com/cilium/cilium/pull/24638#discussion_r1154971360).

christarazi replied:
Ah makes sense. In that case, can we move the `DeepCopy` call to when it's actually needed (as in if we need to perform translation). We should prevent performing a copy of the CNP in the general case (when user is not using CIDRGroupsRefs).

pippolo84 replied:
Actually, it should already be like that.

In case of policy update, we call `translateFromCIDRGroupRefs` only after checking that the cnp has a reference to at least one CIDRGroup in `resolveCIDRGroupRef`:
```
	refs := getCIDRGroupRefs(cnp)
	if len(refs) == 0 {
		return cnp
	}
```

otherwise we return the cnp itself with no further copy.
In case of an update or a deletion of a referenced CIDRGroup, we call `resolveCIDRGroupRef` only after checking that the current cnp actually has a reference to the updated CIDRGroup:

```
		if !hasCIDRGroupRef(cnp, cidrGroup.Name) || cnp.RequiresDerivative() {
			continue
		}
```

In case of a reference being present, we need the double copy because:
- the cache needs to contain the "original" cnp ahead of translation
- `k.updateCiliumNetworkPolicyV2`might modify the `oldRuleCpy` passed in, since the method `.Parse()` it is called on it and that triggers a sanitization of the rules specs: (`*IngressCommonRule) SetAggregatedSelectors()` is an example of a potential write due to the sanitization.

christarazi commented on pkg/policy/cidrgroup/cell.go:
```diff
@@ -0,0 +1,35 @@
+		return ipcache
+	}),
+	cell.Invoke(registerManager),
+)
```
I thought we determined that we don't need this code at all, right? The checkbox from [this](https://github.com/cilium/cilium/pull/24638#issuecomment-1489249279) comment.

christarazi replied:
I'm fairly certain the failures in CI are due to this file. There's a panic inside the code, but we don't all of this code anyway.

christarazi replied:
To reiterate, we don't need this identity allocation code because the identity allocation should happen in the same place as it already does today when we process CIDR blocks in the CNP handler. We don't need to do additional identity allocation because the CIDRGroups are just syntactic sugar on top of CNPs.

pippolo84 replied:
Ok, now I got it, thanks!

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -694,3 +683,31 @@ func cidrGroupRefsToCIDRsSets(cidrGroupRefs []string, cache map[string]*cilium_v
+				found = true
+				break
+			}
+		}
```
This is incorrect. It's looping over `cnp.Spec.Ingress` and not `Specs`. I think this needs to be `cnp.Specs[i].Ingress`.

christarazi replied:
And then ensure that this case is covered by the unit tests below as it doesn't seem to have been caught

pippolo84 replied:
Good catch, I'll add a test to check that, thanks!

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -160,22 +244,25 @@ func (k *K8sWatcher) onUpsertCNP(
 	}
 	if err == nil {
-		cache[key] = cnpCpy
+		// We keep in the cache the version of the cnp ahead of any CIDRGroupRef translation
```
Can we clarify this to say "The CNP cache contains the original form of the CNP (pre-translation of CIDRGroupRef translation)"? I think it also makes more sense to move the comment to where the `cnpCache` is initialized.

christarazi commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1462,7 +1462,7 @@ spec:
                       type: array
                     fromCIDRGroupRefs:
-                      description: FromCIDRGroupRefs is a list of CIDRs which the
+                      description: FromCIDRGroupRef is a list of CIDRs which the
```
Nit: What's the reason to rename it to singular form? I think it was accurate as plural because the user can insert multiple refs in the same `fromCIDRGroupsRefs` field.

pippolo84 replied:
That was my initial thought, but looking at other similar CNP fields (e.g.: `fromCIDR` and `fromCIDRSet`) that are reported as singular despite being slices, I thought to change `FromCIDRGroupRef` too.
I left it as a separate commit for now, if you think that the plural version is better, I can revert it easily.

christarazi replied:
Ok, I see. Better to be consistent then, thanks.

christarazi commented on pkg/k8s/shared_resources.go:
```diff
@@ -52,6 +53,7 @@ type SharedResources struct {
 	Identities                       resource.Resource[*cilium_api_v2.CiliumIdentity]
 	CiliumNetworkPolicies            resource.Resource[*cilium_api_v2.CiliumNetworkPolicy]
 	CiliumClusterwideNetworkPolicies resource.Resource[*cilium_api_v2.CiliumClusterwideNetworkPolicy]
+	CIDRGroup                        resource.Resource[*cilium_api_v2alpha1.CiliumCIDRGroup]
```
Nit: To match the rest of the code above

```suggestion
	CIDRGroups                        resource.Resource[*cilium_api_v2alpha1.CiliumCIDRGroup]
```

christarazi committed:
watchers: Add basic metrics for CCG

Inspired from the CNP watcher.

Co-authored-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +283,98 @@ func (k *K8sWatcher) onDeleteCNP(
 	return err
 }
 
+func (k *K8sWatcher) onUpsertCIDRGroup(
```
Might be cleaner to move all the CIDRGroup functions to a new file (i.e. `cilium_cidr_group.go`) so this file isn't overloaded with too many CIDR group code.

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +283,98 @@ func (k *K8sWatcher) onDeleteCNP(
+		} else {
+			k.K8sEventProcessed(resources.MetricCCNP, resources.MetricUpdate, err == nil)
+			k.K8sEventReceived(k8sAPIGroupCiliumNetworkPolicyV2, resources.MetricCNP, resources.MetricUpdate, true, false)
+		}
```
This is probably misleading because we don't want to increment a metric around K8s event handling inside the CCG logic, because technically there was no {C,}CNP event from K8s. The CCG handler just trigger an internal update for the policies. 

I wonder if we are better off implementing a new metric. Maybe something like `cidr_group_users` which is incremented when a {C,}CNP references a CIDR group?

pippolo84 replied:
Agree, I think a new metric is cleaner and more effective in helping us understand the impact of the CiliumCIDRGroup events.

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +283,98 @@ func (k *K8sWatcher) onDeleteCNP(
+		} else {
+			k.K8sEventProcessed(resources.MetricCCNP, resources.MetricUpdate, err == nil)
+			k.K8sEventReceived(k8sAPIGroupCiliumNetworkPolicyV2, resources.MetricCNP, resources.MetricUpdate, true, false)
+		}
```
Same here regarding the metrics comment above, but the inverse for the delete case, i.e. decrement new metric.

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +283,98 @@ func (k *K8sWatcher) onDeleteCNP(
+
+		translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
+
+		err := k.updateCiliumNetworkPolicyV2(cs, cnpCpy, translatedCNP, initialRecvTime)
```
Inside this call, there's a log line that eventually says "Imported CiliumNetworkPolicy" which is technically not true either because this call is meant for when a new policy is applied or a policy is modified. 

I wonder if we should have a log line for similar for CIDR groups, so at least a user observing their cluster can understand the situation where they have a 1 CNP applied and 1 CCG applied, but see 2 "Imported CiliumNetworkPolicy" msgs. At least they will also see the CIDR group msg too.

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -388,3 +572,147 @@ func reportCNPChangeMetrics(err error) {
+			logfields.K8sAPIVersion:           cnp.TypeMeta.APIVersion,
+			logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
+			logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+			"cidrGroupRefs":                   refs,
```
`logfields.CIDRGroupRef`

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -191,6 +283,98 @@ func (k *K8sWatcher) onDeleteCNP(
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	for key, cnp := range cnpCache {
```
I think it would be good to start measuring the time it takes to go through this loop and perform the CCG translation. It could be useful info for us to measure if the translation needs to be optimized. 

A user can take this new metric subtracted by the CNP processing time (`ccg_translation_time - k8s_event_processed{cnp}`) to see the cost of `resolveCIDRGroupRef()`.

pippolo84 committed:
watchers: Add a metric to count policies referencing a CIDRGroup

Add a gauge metric to count the current number of CNPs and CCNPs that
reference at least a CiliumCIDRGroup object.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
watchers: Add a metric for CIDRGroupRef translation in CNPs

Add a metric to measure the time taken to translate the FromCIDRGroupRef
field in CNPs and CCNPs.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
watchers: Report policy changed event for CCNP too

policy_change_total metric counts all the policies change events and
should take into account Cilium Clusterwide Network Policied events too.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joestringer commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -82,51 +83,143 @@ func (r *ruleImportMetadataCache) get(cnp *types.SlimCNP) (policyImportMetadata,
-	var hasSynced atomic.Bool
-	apiGroup := k8sAPIGroupCiliumNetworkPolicyV2
-	metricLabel := resources.MetricCNP
+	var cnpSynced, ccnpSynced, cidrGroupSynced atomic.Bool
```
I don't understand why CCG needs to be coupled with network policies. IMO they should not be considered a network policy construct, but instead just a concept that "This IP is associated with this name/labels". Then we can also use it for hubble observability and other use cases rather than tying it directly into the policy engine.

joestringer replied:
This is probably my biggest piece of feedback. At a high level, this code injects mutation code into the watcher for all CNP,CCNP objects in order to transform those policies into policies that contain lists of CIDRs corresponding to the groups. There's a few consequences from this, such as strongly coupling the resources together, hiding the cidrgroups from the rest of cilium, requiring additional policy iteration whenever any of these different objects change. I don't know if the "combined handler" is a strict requirement, but that in itself is a red flag to me that introducing this brand new field in policy requires reworking the whole way that policies are ingested into Cilium. That change by itself may or may not make sense, but it should be independently proposed and reviewed IMO.

joestringer replied:
I have a counter-proposal: Decouple CIDRGroup objects from network policy entirely. 

There's two aspects to implementing this feature overall, firstly the `CIDRGroupRef` field in Cilium network policy objects. Secondly, the CIDRGroup ingestion.

The network policy CidrGroupRefs can be translated into a labelselector for `cidrgroup:name=<name>`. This could happen for example as part of `parseToCiliumIngressCommonRule()`. This requires minimal changes to the network policy watchers and doesn't involve introducing any additional rule iteration or calculation when different resources change.

Separately, when updates for CiliumCIDRGroups are received, we can translate that into calls to `ipcache.UpsertLabels(...)`. We'd have some canonical representation of the label such as `{Source: "CIDRGroup",Key:"name",Value:"<name>"}`. For each CIDR in the CIDRGroup, we can inform the ipcache that this CIDR corresponds to this set of labels. Similar to the kube-apiserver entity feature, this will automatically associate these labels with identities for the relevant IP addresses and automatically trigger calculation of any policy impacts whenever the ipcache is informed about this new label association.

I think that this will have a few benefits:
- Simpler implementation/policy handling code
- Cheaper policy parsing/ingestion
- Native hubble visibility

joestringer replied:
Ah that's right, we still have tech debt to address before the ipcache can support this approach: Specifically, if you do `UpsertLabels(10.0.0.0/8, "A")` + `UpsertLabels(10.0.0.1/32, "B")` then the ipcache doesn't currently resolve the overlap between these two. That is to say, rather than considering `10.0.0.1` IP has labels A _and_ B, it would only have B because that's the most specific set of labels. That caveat would need to be addressed if we were to take this approach.

(Consequence would be that it would be possible for two different overlapping CIDRGroupRefs to interfere with one another, ie potentially fail policy matches, if we didn't address this specific issue)

joestringer replied:
I'm still a little concerned about the modularity impacts of adding this much policy-related code into the `pkg/k8s` watcher, but having reviewed the lifecycle impacts of these policy changes, it doesn't look like this will have functional impacts. I also appreciate that there are tests for the core functionality to validate the resource merging functionality. Furthermore I now understand that the motivation behind combining the watchers is to ensure a critical section for updates between the different resource types, to avoid having to introduce some locking between watchers or some other synchronization mechanism. Seems like there's tech debt whichever approach we take on this front, since even with my proposal above we would still need to separately address some other longer-running tech debt. With this current approach, we would sooner or later likely want to revert this in order to better decouple the policy and CIDRGroup resource handling, or at least to avoid the need to place this logic under `pkg/k8s`. At least with the current approach, the vast majority of the logic is self-contained in these two files.

We had some discussions out-of-band about other potential mechanisms to integrate with such as the `statedb` work, definitely an interesting avenue to explore but this is also pretty open-ended. I wouldn't want to expand the scope of this work so much by taking that approach. I briefly also mulled over integrating this functionality into the `TranslateRules()` infrastructure, but the incremental benefit seems low if we aim to change this to an ipcache-based approach in future anyway. Realistically, this code is still relatively self-contained so I anticipate the maintenance burden to be pretty low unless we're planning on significantly refactoring `pkg/k8s/watchers` any time soon. I searched, and I don't see any Modularization issues for this package so it seems like this is a reasonable middle ground for the short to mid term. As long as we have the API right, we at least keep the options on the table for future internal code reworking :+1:

joestringer commented on pkg/k8s/apis/cilium.io/v2alpha1/cidrgroups_types.go:
```diff
@@ -0,0 +1,47 @@
+type CiliumCIDRGroupSpec struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=0
+	CIDRs []api.CIDR `json:"cidrs"`
```
Given that these CIDRs end up only selecting peers outside the cluster, I'd suggest naming something like this:

```suggestion
	ExternalCIDRs []api.CIDR `json:"external-cidrs"`
```

This would make the semantics a bit more clear when they interact with other resources like CNP, since users are less likely to assume that CIDRGroupRefs also select CIDRs for peers inside the cluster.

joestringer replied:
(We could probably elaborate a bit more as well in comments, to make these semantics more obvious for users)

joestringer commented on pkg/policy/api/ingress.go:
```diff
@@ -65,6 +65,15 @@ type IngressCommonRule struct {
+	// This will match on the source IP address of incoming connections.
+	//
+	// +kubebuilder:validation:Optional
+	FromCIDRGroupRef CIDRGroupRefSlice `json:"fromCIDRGroupRefs,omitempty"`
```
Did you consider embedding this under `fromCIDRSet` above so that we don't need to resolve the semantics between the combination of `fromCIDRGroupRefs` + `fromCIDRSet`?

ie suggesting:

```
...
  ingress:
  - fromCIDRSet:
    - cidrGroupRef: "gke-control-plane"
```

The main concern I have here is that under the current proposal, technically someone could write:

```
  ingress:
  - fromCIDRSet:
    - cidrGroupRef: "gke-control-plane"
    fromCIDRGroupRefs:
    - "gke-control-plane"
```

In the above, the natural reading of the single rule is that the IPs must match *both* the `fromCIDRSet` statement and the `fromCIDRGroupRefs`, which probably means that it's the intersection of those two lists(?).. alternatively we would define different semantics for this combination vs. any other combination at this rule layer (see `IngressRule` definition, specifically line 87 onwards). Both of these approaches seem like they should be avoidable.

pippolo84 replied:
Embedding `cidrGroupRef` under `fromCIDRSet` has a couple of ambiguities, too:

- In the CIDRRule, the field CIDR is [required](https://github.com/cilium/cilium/blob/master/pkg/policy/api/cidr.go#L41). Changing that to optional to enable the cidr group reference embedding will introduce the possibility to have empty CIDRRules.
- The Ingress rule sanitization [does not allow](https://github.com/cilium/cilium/blob/master/pkg/policy/api/rule_validation.go#L103) combining `FromCIDR` and `FromCIDRSet`. This is unfortunate, since after the translation of the cidr group reference in `FromCIDRSet` we add a list of CIDRs right in `FromCIDR`.

Regarding the second, I see that there is a `CIDR` Field in the `FromCIDRSet` rule, but a single cidr group could hold more than one CIDR. Using that, the translation of a single "fromCIDRSet/cidrGroupRef" rule will result in multiple "fromCIDRSet/CIDR" rules, one for each CIDR in the cilium cidr group.

joestringer replied:
Hmm I see. Technically it would be possible to support this with `oneOf` semantics, but it looks like the kube-builder code generator we use doesn't support that statement.

One alternative would be to move this code later in the process to a post-parsing stage so that the effects of this translation do not need to be subject to the same checks as are being enforced at the API layer here.

Incidentally, how does the 'there must be at least one CIDR' check pass today with an empty CIDR group?

Another option may be to not change the location of the field, but instead just add validation code (+ API comments) to prevent the combination of top level statements that I had raised above. The concern I had is around the combination of specific statements, but we could define that if there is a CIDRGroupRef then the user cannot also specify CIDRs/CIDRSets. Then it's "just" up to the internal logic to skip that API check iff the CIDRs are generated.

joestringer replied:
Another thought, we could just change the CIDR field to optional at the API layer. Then internally when parsing / validating the policy we enforce that at least one of the fields must be present.

Or another similar variation, just mark it optional. If the CIDR and CIDRGroupRef are empty, we decide some semantics for that (match all or match none). Not a huge fan of this one since we have to commit to another policy semantics change.

pippolo84 replied:
After an out-of-band discussion and some experimentation, we decided to embed the field `fromCidrGroup` in the `fromCIDRSet` Ingress rule as suggested. This is possible since out fork of kubebuilder supports the `oneOf semantic.
Regarding the translation, for each referenced cidr group, the CNP/CCNP/CCG handler generates additional `fromCIDRSet` objects containing the CIDRs belonging to the groups.

joestringer commented on install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml:
(Unrelated to this file)

I don't see an update for `install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml`. How is the new CRD registered?

pippolo84 replied:
Good catch, I've accidentally deleted the change in the last rebase, I'll fix it.

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,281 @@
+	// If the referenced CiliumCIDRGroup is empty, and no other rule is
+	// specified in the Ingress part of the policy, we must set the entire
+	// Ingress field to nil, otherwise the policy will deny all ingress traffic.
+	removeEmptyIngresses(translated)
```
The network policy says "set a default deny, and allow all CIDRs in the CIDRGroup". If the CIDRGroup is empty, then that *should* deny all traffic. I think that the logic here where empty cidrgroups change the default-deny posture is a strange semantic in the policy.

I suggest removing this whole `removeEmptyIngresses()` aspect of the solution.

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1461,6 +1461,20 @@ spec:
                         type: string
                       type: array
+                    fromCIDRGroupRefs:
+                      description: FromCIDRGroupRef is a list of CIDRs which the endpoint
```
```suggestion
                      description: FromCIDRGroupRef is a list of CIDRs that the endpoint,
```

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1461,6 +1461,20 @@ spec:
                       type: array
+                    fromCIDRGroupRefs:
+                      description: FromCIDRGroupRef is a list of CIDRs which the endpoint
+                        subject to the rule is allowed to receive connections from
```
```suggestion
                        subject to the rule, can receive connections from
```

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1461,6 +1461,20 @@ spec:
+                    fromCIDRGroupRefs:
+                      description: FromCIDRGroupRef is a list of CIDRs which the endpoint
+                        subject to the rule is allowed to receive connections from
+                        in addition to FromEndpoints, along with a list of subnets
```
I'm not sure what is meant by "along with a list of subnets contained within their corresponding IP block from which traffic should not be allowed". Isn't cidrgroup just a list of CIDRs? Does it have a more complicated structure that I'm not understanding? If my understanding is correct then I think this sentance is unnecessary.

```suggestion
                        (in addition to FromEndpoints).
```

pippolo84 replied:
No, you're right, it's just a copy and paste error from `FromCIDRSet` doc. I'll fix it.

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1461,6 +1461,20 @@ spec:
+                        subject to the rule is allowed to receive connections from
+                        in addition to FromEndpoints, along with a list of subnets
+                        contained within their corresponding IP block from which traffic
+                        should not be allowed. This will match on the source IP address
```
```suggestion
                        This will match on the source IP address of incoming traffic.
```

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1461,6 +1461,20 @@ spec:
+                        in addition to FromEndpoints, along with a list of subnets
+                        contained within their corresponding IP block from which traffic
+                        should not be allowed. This will match on the source IP address
+                        of incoming connections.
```
```suggestion
```

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -2160,6 +2174,20 @@ spec:
                         type: string
                       type: array
+                    fromCIDRGroupRefs:
+                      description: FromCIDRGroupRef is a list of CIDRs which the endpoint
```
All of my suggestions above should apply here as well, except that this should say, 

"FromCIDRGroupRef is a list of CIDRs that the endpoint, subject to the rule, **cannot** receive connections from (in addition to FromEndpoints). This will match on the source IP address of incoming connections.

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -3958,6 +3986,20 @@ spec:
                           pattern: ^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/([0-9]|[1-2][0-9]|3[0-2])$|^s*((([0-9A-Fa-f]{1,4}:){7}(:|([0-9A-Fa-f]{1,4})))|(([0-9A-Fa-f]{1,4}:){6}:([0-9A-Fa-f]{1,4})?)|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){0,1}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){0,2}):([0-9A-Fa-f]{1,4})?))|(([ ...
                           type: string
                         type: array
+                      fromCIDRGroupRefs:
```
ditto "can receive connections from"

nathanjsweet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -4667,6 +4709,20 @@ spec:
                           type: string
                         type: array
+                      fromCIDRGroupRefs:
+                        description: FromCIDRGroupRef is a list of CIDRs which the
```
ditto "cannot receive connections from"

nathanjsweet commented on pkg/policy/api/ingress.go:
```diff
@@ -65,6 +65,15 @@ type IngressCommonRule struct {
 	// +kubebuilder:validation:Optional
 	FromCIDRSet CIDRRuleSlice `json:"fromCIDRSet,omitempty"`
 
+	// FromCIDRGroupRef is a list of CIDRs which the endpoint subject to the
```
This one is a little more complicated to document, as it is an embedded type used for both `Ingress` and `IngressDeny`. I would recommend something like:

"FromCIDRGroupRef is a list of CIDRs that the endpoint, subject to the rule, can or cannot receive connections from (in addition to FromEndpoints). This will match on the source IP address of incoming connections"

pippolo84 replied:
Changing this will also update the yaml files above after running `make generate-k8s-api`. I think this sentence is fine for both `Ingress` and `IngressDeny`. :+1:

nathanjsweet reviewed (changes requested):
Some docs things.

nathanjsweet reviewed (approved):
LGTM!

joestringer commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -82,51 +83,143 @@ func (r *ruleImportMetadataCache) get(cnp *types.SlimCNP) (policyImportMetadata,
+		k8sAPIGroupCiliumClusterwideNetworkPolicyV2,
+	)
+	k.k8sAPIGroups.AddAPI(k8sAPIGroupCiliumClusterwideNetworkPolicyV2)
 }
```
It looks like we need to add a similar set of `blockWaitGroupToSyncResources()`, `AddAPI()` calls for the CIDRGroup type?

Consequence could be that we ingest one network policy with references to various CIDRGRoups, but the CIDRGroups are not yet synchronized from the apiserver => Means that when we process the policy, the BPF policymaps will be programmed with only a subset of the CIDRGroup CIDRs. That could result in temporary drops during startup.

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,281 @@
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	for key, cnp := range cnpCache {
+		if !hasCIDRGroupRef(cnp, cidrGroup.Name) || cnp.RequiresDerivative() {
```
nit: I'm not sure I understand the logical impact of `RequiresDerivative()`, which seems to boil down to "Does this policy have a ToGroups statement?". That check seems to only apply to egress policies but this is an ingress policy feature, and just generally the notion of whether there is some such statement _somewhere_ in the CNP governing whether _this_ feature should work doesn't make sense to me. Is there some background to this that I'm missing?

```suggestion
		if !hasCIDRGroupRef(cnp, cidrGroup.Name) {
```

pippolo84 replied:
No, I agree that it has no impact on the Ingress part of the policy. I'm gonna remove the check for `RequiresDerivative()`.

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,281 @@
+	return err
+}
+
+func (k *K8sWatcher) onDeleteCIDRGroup(
```
nit: I see a few minor differences between `onDeleteCIDRGroup()` and `onUpsertCIDRGroup()`, but in the end they both translate into calls to update the network policy for each CNP that selects this CIDRGroup. Is there any reason to have two different functions or could they be the same handler?

pippolo84 replied:
Apart from the differences in cache management and metrics reporting the code is equal, I moved it in a separate function to be reused. :+1:

joestringer commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -155,36 +248,54 @@ func (k *K8sWatcher) onUpsertCNP(
 }
 
-func (k *K8sWatcher) onDeleteCNP(
+func (k *K8sWatcher) onDelete(
```
comment, no action necessary: I found it a bit odd that the delete case didn't need to do any resolution of the CIDRGroupRefs internally in the way that the upsert case does. However, this is correct since any deletion ends up doing an `api.Rule` (~= `.Spec`) lookup in `policy.Repository` for every rule with the CNP's labels, and deletes those rules by label. Therefore, it doesn't matter what CIDRGroupRefs are present in the deleted CNP since the entire rules will be deleted down at the daemon policy layer. :+1: 

@christarazi and I also looked into similar cases for Upsert where one policy exists, then we update it with additional rules, or update it to delete rules. The logic is similar in that case, basically the `daemon.policyAdd()` function grabs the `policy.Mutex`, deletes all of the old rules by CNP labels, adds all the new rules, (calculates identity refcount impacts), and releases the policy Repository mutex. So there should be no lifecycle implications for those cases.

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -827,6 +847,26 @@ func CreateConfiguration(metricsEnabled []string) (Configuration, []prometheus.C
+				Namespace: Namespace,
+				Name:      "cidr_group_translation_time_stats_seconds",
+				Help:      "CIDRGroup translation time stats",
+			}, []string{LabelPolicyNamespace, LabelPolicyName})
```
Hmm, I don't think we should be using policy names in metrics since that's user input and the cardinality can reach very high.

pippolo84 replied:
Fixed with a histogram without additional labels.

joestringer reviewed (changes requested):
Mainly just one more question on top of my previous feedback from lifecycle review. I took a fresh look at the main watcher code and the the metrics. It sounded like Chris had already reviewed much of the details of `pkg/k8s/watchers/cilium_cidr_group.go` and the corresponding test code, so I skipped over the details there.

pippolo84 committed:
metrics: Add NoOpHistogram variable

The NoOp* variables are used in pkg/metrics to define metrics that can
be disabled.  All of those metrics are grouped together as a slice of
prometheus.Collector and then registered to the prometheus register.

To add a metric of type prometheus.Histogram that can also be disabled,
a type that satisfies both `prometheus.Collector` and
`prometheus.Observer` is needed, so the NoOpHistogram of type
prometheus.histogram is added.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
policy: Add cidrGroupRef field in api.CIDRRule

Embed `cidrGroupRef` field in api.CIDRRule to reference CiliumCIDRGroup
objects in Cilium Network Policies.  The field can be used to allow
ingress traffic from a set of CIDRs listed in the referenced
CiliumCIDRGroup object.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
nit, follow up: These files should probably be in codeowners with ownership under @cilium/sig-policy .

pippolo84 replied:
Assigned `cilium_network_policy.go`, `cilium_cidr_group.go` and `cilium_cidr_group_test.go` to sig-policy.

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,253 @@
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroup.Name, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, action, err == nil)
```
Is this supposed to be counted when the event is processed and determined to make no changes? Currently it's skipped from the `return nil` case above.

pippolo84 replied:
It's the same logic we are already using for [CNP/CCNP](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/cilium_network_policy.go#L143-L173).
I think it is better to keep it consistent.
Also, I think it is more appropriate to not count "spurious" events that do not need any processing.

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,253 @@
+	err := k.updateCIDRGroupRefPolicies(cidrGroupName, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, resources.MetricDelete, err == nil)
+	k.K8sEventReceived(apiGroup, metricLabel, resources.MetricDelete, true, true)
```
The last two parameters here seem a little misleading for `valid`, `equal` given that parsing for instance could fail, or maybe the cidr group ref couldn't be found, etc.

pippolo84 replied:
Actually, it is the same logic we are already using for CNP/CCNP: [[1]](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/cilium_network_policy.go#L140), [[2]](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/cilium_network_policy.go#L173) and [[3]](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/cilium_network_policy.go#L188-L189). Looking at the existent code, it is possible to have errors while updating or deleting CNPs, too.

I agree that it is questionable, but maybe it is like that because the initial intention was to report info about the event reception and not its processing, which is reported by `k.K8sEventProcessed()`?
Maybe we should reconsider the way we are using the labels for all of those metrics (CNP, CCNP and CCG) in a follow up PR?

joestringer replied:
Yep, that seems reasonable. Maybe file an issue with the details and we can follow up separately?

pippolo84 replied:
Opened a related issue [here](https://github.com/cilium/cilium/issues/24873)

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,253 @@
+			logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
+			logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+			logfields.CIDRGroupRefs:           refs,
+		}).WithError(err).Warning("unable to translate all cidr groups to cidrs")
```
nit, I think that errors are supposed to be all lower case since they're concatenated with other strings, but messages should start with a capital as though it's a sentence (this is at least how it seems like most messages are written in the codebase so far, from a quick glance at grep output):

```suggestion
		}).WithError(err).Warning("Unable to translate all cidr groups to cidrs")
```

joestringer commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -421,7 +422,7 @@ var ciliumResourceToGroupMapping = map[string]watcherInfo{
 	synced.CRDResourceName(v2alpha1.LBIPPoolName): {skip, ""}, // Handled in LB IPAM
 	synced.CRDResourceName(v2alpha1.CNCName):      {skip, ""}, // Handled by init directly
-
+	synced.CRDResourceName(v2alpha1.CCGName):      {skip, ""}, // Handled by CNP
```
nit: Why is this marked as `skip` whereas both `CNPName` and `CCNPName` have entries here? 

```suggestion
	synced.CRDResourceName(v2alpha1.CCGName):      {start, k8sAPIGroupCiliumCIDRGroupV2Alpha1},
```

pippolo84 replied:
Good catch, this was a leftover before the change that introduced the `sync.Once` for the shared CNP/CCNP/CCG handler. I'll fix it, thanks :+1:

joestringer commented on pkg/metrics/interfaces.go:
```diff
@@ -37,7 +37,7 @@ var (
 	NoOpCounter     prometheus.Counter     = &counter{NoOpMetric, NoOpCollector}
 	NoOpCounterVec  CounterVec             = &counterVec{NoOpCollector}
-	NoOpObserver    prometheus.Observer    = &observer{}
+	NoOpObserver    prometheus.Histogram   = &observer{NoOpCollector}
```
Two questions:
- Why is this necessary?
- Why continue to call the `NoOpObserver` as an observer if it's trying to implement a different interface? (for example, why not have a new type `NoOpHistogram` for this purpose)?

joestringer replied:
Maybe another option, why not just set `CIDRGroupTranslationTimeStats` to `NoOpCollector`?

pippolo84 replied:
We need a type that satisfies both `prometheus.Collector` and `prometheus.Observer`. The first because [we group](https://github.com/cilium/cilium/blob/master/pkg/metrics/metrics.go#L687) all metrics together in a slice of `prometheus.Collector`. The second because the `prometheus.Observer` interface contains the `Observe` method we use to report the metric in `cilium_cidr_group.go`.

Probably it is better to add a new type called `NoOpHistogram` to avoid confusion with the name. I'll fix it.

joestringer commented on pkg/policy/api/cidr.go:
```diff
@@ -50,6 +50,13 @@ type CIDRRule struct {
 	// +kubebuilder:validation:Optional
 	ExceptCIDRs []CIDR `json:"except,omitempty"`
 
+	// CIDRGroupRef is a reference to a CiliumCIDRGroup object.
```
nit: I would assume that the `OneOf` fields should all show up near the top so it's easier to scan through and see which options are part of the `OneOf` constraint, then all optional fields follow. Is there a reason this is put in the middle?

pippolo84 replied:
I thought to keep the field `ExceptCIDRs` close to `CIDRs` because logically tied to it, but not to `CIDRGroupRef`. At the moment, the translation does not support both `CIDRGroupRef` and `ExceptCIDRs`.
Probably it is better to move `ExceptCIDR` below but also add a comment to point out that specifying something there is not supported when using `CIDRGroupRef`.

joestringer replied:
Is that restriction documented and enforced on the API?

pippolo84 replied:
I documented it (I'm about to push a new revision of the code), but not enforced.
Regarding the enforcement, do you mean something at the kubebuilder validation level or something in the cidr_group handler? For the former I am not aware of anything like that in kubebuilder validation semantic (but I can take a deeper look), for the latter what should we do in that case? Maybe stop the translation and return an error? Or simply ignore the fromCIDRSet element and go ahead?

joestringer replied:
I'm not familiar enough with kube-builder to know if it could enforce that constraint.

kube-builder would be preferred if possible, otherwise we have to make these decisions about exactly how to handle it. As an incremental limitation, as long as the policy still puts the endpoint into default deny, I think it should be okay.. probably that means log+skip the element and continue.

Is there a technical reason not to support it? We could just calculate the set of CIDRs from the group and then subtract these CIDRs right? I think there are some CIDR subtraction helper functions in the tree already :thinking: Maybe another item for https://github.com/cilium/cilium/issues/24801.

pippolo84 replied:
Actually, I realize that support for `ExceptCIDRs` can be easily implemented simply copying the field to the translated rules. I added it alongside some unit testing.

tommyp1ckles commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumcidrgroups.yaml:
```diff
@@ -0,0 +1,66 @@
+  - name: v2alpha1
+    schema:
+      openAPIV3Schema:
+        description: 'CiliumCIDRGroup is a list of external CIDRs (i.e.: CIDRs selecting
```
nit: `i.e.:` -> `i.e:`

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -626,6 +643,8 @@ func DefaultMetrics() map[string]struct{} {
 		Namespace + "_policy_change_total":                                           {},
 		Namespace + "_policy_endpoint_enforcement_status":                            {},
 		Namespace + "_policy_implementation_delay":                                   {},
+		Namespace + "_cidr_group_translation_time_stats_seconds":                     {},
```
nit: I wonder whether we really need to know about this by default, whether it's expensive enough to need to know about it in production at all, etc. This will add a few extra buckets to the metrics if it's in use, but seems like it would be most useful if we already suspect a performance issue with the CIDR Group handling logic. Maybe we should just keep the metric but drop this line so it's not auto-enabled?

joestringer replied:
@tommyp1ckles raised an interesting adverse scenario for this below which might help inform this decision: 
https://github.com/cilium/cilium/pull/24638#discussion_r1162183486

pippolo84 replied:
As said in https://github.com/cilium/cilium/pull/24638#discussion_r1162183486 I don't think we'll face many troubles regarding performance, at least not in common scenarios. So I'm inclined to say that removing the metric default-enable is reasonable.

tommyp1ckles commented on pkg/k8s/apis/cilium.io/v2alpha1/cidrgroups_types.go:
```diff
@@ -0,0 +1,50 @@
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=0
+	ExternalCIDRs []api.CIDR `json:"external-cidrs"`
```
For K8s CRDs, isn't the convention to use camelCase for the json/yaml serialization?

joestringer replied:
(This is more common, I might've misled here since I requested the rename. I'm happy with ExternalCIDRs instead)

pippolo84 replied:
Should I go for `externalCidrs` or `externalCIDRs`?

joestringer replied:
CIDR is an acronym, so `externalCIDRs` seems more appropriate. This would also match the precedent with `toFQDNs`.

tommyp1ckles commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,227 @@
+) error {
+	delete(cidrGroupCache, cidrGroupName)
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroupName, cidrGroupCache, cnpCache, cs)
```
nit: can just be `return k.updateCIDRGroupRefPolicies(cidrGroupName, cidrGroupCache, cnpCache, cs)`

tommyp1ckles commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,227 @@
+	cs client.Clientset,
+) error {
+	var errs []error
+	for key, cnp := range cnpCache {
```
Why do we not do anything for CCNP cache, are CCNP refs not supported?

gandro replied:
Since I also wondered about this during the review. It seems like we're unifying CNP and CCNP handling, so despite the name, `cnpCache` will contain CCNPs. So CCNP upserts share the same upsert path as CNPs here:

https://github.com/cilium/cilium/blob/9d9b71aca65ab9c45f898f9eb4729c92f6b8051c/pkg/k8s/watchers/cilium_network_policy.go#L160

Which then adds the slim.CNP (which can be a CCNP) here:
https://github.com/cilium/cilium/blob/9d9b71aca65ab9c45f898f9eb4729c92f6b8051c/pkg/k8s/watchers/cilium_network_policy.go#L266

Since `resource.Key` contains the namespace, the key should also be unique between CNPs and CCNPs with the same name.

But I'll also let Fabio confirm this.

christarazi replied:
There was some discussion with Joe and I about making a separate refactoring commit which is clearly titled as "Unify CNP and CCNP handling in watchers" which would've made this very clear and also be the place to explain why it's necessary. @pippolo84 if this is not too much work, it might be good to have this.

pippolo84 replied:
Yep, np, I'm gonna split the commit.

joestringer commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -379,6 +479,11 @@ func (k *K8sWatcher) updateCiliumNetworkPolicyV2AnnotationsOnly(ciliumNPClient c
 
 }
 
+func (k *K8sWatcher) registerResourceWithSyncFn(ctx context.Context, resource string, syncFn func() bool) {
+	k.blockWaitGroupToSyncResources(ctx.Done(), nil, syncFn, resource)
+	k.k8sAPIGroups.AddAPI(resource)
+}
+
```
nit: This should probably exist in the generic `pkg/k8s/watchers/watcher.go`, or at some point, not exist at all because it seems like every watcher has separate calls to these functions, which is probably a sign it could be hooked all into a single registration function that's automatically invoked by all watchers. But that's out-of-scope for this PR.

christarazi replied:
Added to https://github.com/cilium/cilium/issues/24801

squeed replied:
an update: this is no longer a factor; the file in question is gone :-)

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,253 @@
+			logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+			logfields.CIDRGroupRefs:           refs,
+		}).WithError(err).Warning("unable to translate all cidr groups to cidrs")
+	}
```
(I note that if there are valid references to cidr groups and invalid references to cidr groups, then the invalid ones will be ignored and the valid ones will be used. This makes sense to me, as long as the invalid ones ensure a default deny posture)

pippolo84 replied:
Yes, the default deny posture is ensured in case the `Ingress` section ends up to be empty after the translation (that is, if there are only CIDRGroupRef rules and all of them ends up referencing empty or non-existent sets).

joestringer commented on pkg/policy/api/cidr.go:
```diff
@@ -50,6 +50,13 @@ type CIDRRule struct {
 	ExceptCIDRs []CIDR `json:"except,omitempty"`
 
+	// CIDRGroupRef is a reference to a CiliumCIDRGroup object.
+	// A CiliumCIDRGroup contains a list of CIDRs that the endpoint, subject to
+	// the rule, can (Ingress) or cannot (IngressDeny) receive connections from.
```
Are cidrGroupRefs case-sensitive or case-insensitive? We should probably define that in the API and also ensure the parsing logic accounts for this.

pippolo84 replied:
They are case sensitive.
Since we are using the names to identify the CiliumCIDRGroup objects, they must be lowercase RFC1123 strings. I'll change the validation pattern for `CIDRGroupRef` type to enforce this.

tommyp1ckles commented on pkg/metrics/metrics.go:
```diff
@@ -827,6 +841,16 @@ func CreateConfiguration(metricsEnabled []string) (Configuration, []prometheus.C
 			collectors = append(collectors, PolicyImplementationDelay)
 			c.PolicyImplementationDelayEnabled = true
 
+		case Namespace + "_cidr_group_translation_time_stats_seconds":
```
should this maybe be  `_cidrgroup_`or `_cg_` (not sure, seems like cidrgroup is a proper noun so it should be grouped)?

pippolo84 replied:
Let's go for `_cidrgroup_`, I'll change it :+1:

joestringer commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,253 @@
+			}
+		}
+		if !found {
+			errs = append(errs, fmt.Errorf("cidr group %s not found, skipping translation", cidrGroupRef))
```
nit: The user input should probably be quoted to make it clear that this is user input and not part of the static message.
```suggestion
			errs = append(errs, fmt.Errorf("cidr group %q not found, skipping translation", cidrGroupRef))
```

tommyp1ckles commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,227 @@
+			}
+		}
+	}
+	for _, spec := range cnp.Specs {
```
nit: might be able to DRY this up with something like:
```go
for _, spec := range append(cnp.Specs, cnp.Spec) {
    if spec == nil {
        continue
    }
    ...
}
```

pippolo84 replied:
Nice one, thanks!

joestringer reviewed (commented):
One more query on the API (case-sensitivity for CIDRGroups), and a scattering of minor feedback, main one of note was around some of the metrics handling for k8s objects and whether to enable some of the new metrics by defaults.

tommyp1ckles commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,227 @@
+			if cidrGroupName == cidrGroupRef {
+				cidrs := make([]api.CIDR, 0, len(cidrGroup.Spec.ExternalCIDRs))
+				for _, cidr := range cidrGroup.Spec.ExternalCIDRs {
+					cidrs = append(cidrs, api.CIDR(cidr))
```
Maybe I missed it - is there validation that these are indeed cidrs somewhere?

pippolo84 replied:
Yes, there is. `cidrGroup.Spec.ExternalCIDRs` is declared as `[]api.CIDR` and `api.CIDR` has already a kubebuilder validation pattern set up.

tommyp1ckles commented on pkg/k8s/watchers/cilium_cidr_group.go:
```diff
@@ -0,0 +1,253 @@
+	cs client.Clientset,
+) error {
+	var errs []error
+	for key, cnp := range cnpCache {
```
Could there be any performance concerns with a cluster with a lot of CNP/Refs doing many updates (Particularly, because this now serializes CNP/CG events into one loop). We're also duplicating CNP watch.Store with this map.

Would it make any sense to store a ref index onto CNPs?  Or perhaps we could use an indexed Store similar to CEP/CES?

joestringer replied:
^ Might be a good case for some sort of Go benchmark to get a sense for the performance of it. We do have the metrics at least to be able to measure the impact.

tommyp1ckles replied:
Yeah don't want to encourage premature optimization, but it might be worth due to it being tied up with the CNP update path now.

pippolo84 replied:
I agree that having a benchmark will be useful. The scenario depicted by Tom might happen, but keep in mind that the `resource.Resource` type works like a typed SharedInformer + workqueue. So, in case of a high number of events where every single one cannot be served immediately because of the serialization, the burst will now be "collapsed" together. Even if we are merging the handlers, I think that this is possibly more efficient than what we were doing just before https://github.com/cilium/cilium/pull/24509, when only an informer was used.
Even if it is not possible to anticipate anything, I'd say that the impact of this PR shouldn't be bad, especially for a reasonable number of cidr group references.
The long term solution is to rely on the statedb or something similar as discussed [here](https://github.com/cilium/cilium/pull/24638#discussion_r1160285173).

christarazi replied:
We discussed this item a bit more:

We determined policy implementation delay is more concerning in terms of impact of this code instead of the CNP translation time code. We don't expect the CNP translation time to have a significant impact here. Given the state of the controlplane tests and that fact that doesn't yet support endpoint regeneration due to lack of datapath mocking, we can't yet devise controlplane test to stress this code without doing it manually. So the idea would be to create the resources manually and observe policy implementation delay metric. 

Memory is not the concern because the policies and the CIDRs within the policy are at maximum ~1KB and 1000 of these policies are only ~1MB. The concern would be more CPU.

The impact to user if we have a performance issue here is increased delay for policy recalculation.

I added this as something to followup on in https://github.com/cilium/cilium/issues/24801.

gandro reviewed (approved):
Awesome work! I did not focus on the CRD design itself (since more knowledgeable reviewers seem to have already done that), the overall logic itself looks sound to me, assuming some of the outstanding comments will be resolved.

pippolo84 committed:
CODEOWNERS: Define ownership for CNP/CCNP/CCG handler code

Assign the following files:

- k8s/pkg/watcher/cilium_network_policy.go
- k8s/pkg/watcher/cilium_cidr_group.go
- k8s/pkg/watcher/cilium_cidr_group_test.go

containing the shared handler for CNP, CCNP and CCG objects to
sig-policy group.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -216,6 +216,12 @@ const (
+	LabelPolicyNamespace = "policyNamespace"
+
+	// LabelPolicyName is the label for the name of the policy that references a CIDRGroup
+	LabelPolicyName = "policyName"
```
I think that policy names like this are the sort of thing we should really try to avoid putting in metrics. This is user input with basically no bound on the cardinality. This can cause performance and $$$ problems for users when these metrics are triggered.

Namespace is also a bit sketchy for similar reasons, but more reasonable than policy names.

pippolo84 replied:
Same as https://github.com/cilium/cilium/pull/24638#discussion_r1164771220

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -216,6 +216,12 @@ const (
 	LabelAddressTypeSecondary = "secondary"
+
+	// LabelPolicyNamespace is the label for the namespace of the policy that references a CIDRGroup
+	LabelPolicyNamespace = "policyNamespace"
```
For that matter, I'm not seeing these labels getting used anywhere anyway. If we're not using them, then this code should be removed. Or they should be used when observing the metrics.

pippolo84 replied:
It's a leftover from the previous revision, I'll remove it.

joestringer commented on pkg/policy/api/cidr.go:
```diff
@@ -171,3 +179,11 @@ func addrsToCIDRRules(addrs []netip.Addr) []CIDRRule {
 }
+
+// +kubebuilder:validation:MaxLength=253
+// +kubebuilder:validation:Pattern=`^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
```
How many escapes do we need for the `.` in kubebuilder?

pippolo84 replied:
This is the validation pattern for `dns1123SubdomainFmt`. It is used [here](https://github.com/cilium/cilium/blob/master/pkg/k8s/slim/k8s/apis/meta/v1/types.go#L337) too, among the others.

joestringer reviewed (commented):
Couple more nits on unused metrics labels and whether we should use them.

joestringer commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -67,6 +67,7 @@ const (
 	k8sAPIGroupNetworkingV1Core                 = "networking.k8s.io/v1::NetworkPolicy"
 	k8sAPIGroupCiliumNetworkPolicyV2            = "cilium/v2::CiliumNetworkPolicy"
 	k8sAPIGroupCiliumClusterwideNetworkPolicyV2 = "cilium/v2::CiliumClusterwideNetworkPolicy"
+	k8sAPIGroupCiliumCIDRGroupV2Alpha1          = "cilium/v2alpha1::CiliumCIDRGroup"
```
nit: This ended up in the refactor commit, not the CIDRGroup logic commit.

joestringer reviewed (approved):
LGTM, just one super minor nit. Great work. The splitting up of the logic for the refactor vs. the new feature also made the PR much easier to reason about :clap:

christarazi commented on pkg/metrics/metrics.go:
```diff
@@ -630,6 +641,7 @@ func DefaultMetrics() map[string]struct{} {
 		Namespace + "_policy_change_total":                                           {},
 		Namespace + "_policy_endpoint_enforcement_status":                            {},
 		Namespace + "_policy_implementation_delay":                                   {},
+		Namespace + "_cidr_group_policies":                                           {},
```
This is leading to 

```
level=warning msg="Metric does not exist, skipping" metric=cilium_cidr_group_policies
```

in CI.

Looks like we have just a typo, `_cidrgroup_policies` vs `_cidr_group_policies`.

christarazi commented on pkg/k8s/watchers/cilium_network_policy.go:
```diff
@@ -159,6 +159,7 @@ func (k *K8sWatcher) ciliumNetworkPoliciesInit(ctx context.Context, cs client.Cl
 				case resource.Delete:
 					err = k.onDelete(slimCNP, cnpCache, event.Key, k8sAPIGroupCiliumClusterwideNetworkPolicyV2, resources.MetricCCNP)
 				}
+				reportCNPChangeMetrics(err)
```
I guess this was a "bug" from the original code and we are fixing it now because the refactor exposed it, it's fine IMO. Typically, we'd want to separate this out into a separate PR, but I think it's pretty minor given that this was probably missing in the code for some time? Anyway, I don't think it's worth too much to look into this.

pippolo84 replied:
Yes, it is. Being so small, I just added a separate commit in this PR to add the metric reporting also for CCNP.

christarazi reviewed (approved):
Fantastic work!
